### ISSUE TITLE
Support go 1.18 

### ISSUE THREAD

**Issue by camilamacedo86** [kubernetes-sigs/kubebuilder#2559]
### What do you want to happen?

Make Kubebuilder supports Golang 1.18 

**Following the main places that need to be changed to achieve this goal:**

- Bump go.mod scaffold for go/v3 scaffolds 
- Also, ensure that we bump the go version in the GitHub Actions and CI. 
- Verify any other places that we should do this update and ensure that all will be properly done. 
- Ensure that you make generate before pushing the PR with the changes. 

**Following what was identified that would also be required to be done to achieve this goal**

- [x] Update for go/v3 controller-runtime release with k8s 1.24 on Kubebuilder : https://github.com/kubernetes-sigs/kubebuilder/issues/2681
- [x] Update for go/v3 controller-tools with a release using k8s 1.24 
- [x] Update K8s version to 1.24
- [x] Update prow tests to run the test with go 1.18: PR: https://github.com/kubernetes/test-infra/pull/25775
- [x] Add prow test for k8s 1.24
- [x] To build the Kubebuilder-tools for k8s 1.24

**We should NOT**
- update go/v2 since we cannot upgrade the deps used on it, and they will not be compatible with go 1.18. 
- we should not bump the go 1.18 on the Kubebuilder go.mod for now because it is used as lib: (To allow people to consume it and we should adopt the same approach from C+R and controller-tools. More info:https://kubernetes.slack.com/archives/C02MRBMN00Z/p1652178727902549?thread_ts=1652177106.531319&cid=C02MRBMN00Z)

**Comment by AvineshTripathi** [comment]
@camilamacedo86 i would like to do this, pretty new to kubebuilder so i was wondering if this could be a issue to start with

**Comment by camilamacedo86** [comment]
I think that is good @AvineshTripathi.
Feel free to ping me if you need.

**Comment by AvineshTripathi** [comment]
Looking at previous bump of go version in kubebuilder https://github.com/kubernetes-sigs/kubebuilder/pull/2182/commits/15e3ef35aafacb8a1a1797720470e66ff7745cfc for 1.15->1.16 all the dockerfiles present in the docs/books/src were too updated to new version are we suppose to do the same this time @camilamacedo86

**Comment by abeer91** [comment]
Not sure if this needs a whole other issue, I can create one if needed. Currently, the version of kube builder available through home brew depends on golang 1.18 , and it doesn't work with the same create-api errors reported.

https://formulae.brew.sh/formula/kubebuilder

**Comment by AlmogBaku** [comment]
We probably need to have a new release to solve it. Merging it will allow you to temporary solve it by using the `head` version

**Comment by camilamacedo86** [comment]
Hi @abeer91, 

the brew formula is not kept maintained by us.
If it is using go 1.18 that is wrong since KB still not supporting this version.

**Comment by abeer91** [comment]
Thanks @camilamacedo86  , do you know who maintains this?

**Comment by AlmogBaku** [comment]
@abeer91, you should probably open an issue in the [brew-core](https://github.com/Homebrew/homebrew-core) repo

**Comment by viveksahu26** [comment]
Hey @camilamacedo86 , I found similar to above issue in one of  [KataCoda](https://www.katacoda.com/javajon/courses/kubernetes-extensibility/kubebuilder)  scenario, `Kubernetes Extensibility: Develop an Operator with Kubebuilder`. 
```
controlplane $ go version
go version go1.18.1 linux/amd64
controlplane $ kubebuilder init --domain d2iq.com
Error: failed to initialize project: unable to run pre-scaffold tasks of "base.go.kubebuilder.io/v3": target directory is not empty (only go.mod, go.sum, files and directories with the prefix ".", files with the suffix ".md" or capitalized files name are allowed); found existing file "go1.18.1.linux-amd64.tar.gz"
Usage:
  kubebuilder init [flags]

Examples:
  # Initialize a new project with your domain and name in copyright
  kubebuilder init --plugins go/v3 --domain example.org --owner "Your name"

  # Initialize a new project defining a specific project version
  kubebuilder init --plugins go/v3 --project-version 3


Flags:
      --component-config         create a versioned ComponentConfig file, may be 'true' or 'false'
      --domain string            domain for groups (default "my.domain")
      --fetch-deps               ensure dependencies are downloaded (default true)
  -h, --help                     help for init
      --license string           license to use to boilerplate, may be one of 'apache2', 'none' (default "apache2")
      --owner string             owner to add to the copyright
      --project-name string      name of this project
      --project-version string   project version (default "3")
      --repo string              name to use for go module (e.g., github.com/user/repo), defaults to the go package of the current working directory.
      --skip-go-version-check    if specified, skip checking the Go version

Global Flags:
      --plugins strings   plugin keys to be used for this subcommand execution

2022/04/16 22:43:28 failed to initialize project: unable to run pre-scaffold tasks of "base.go.kubebuilder.io/v3": target directory is not empty (only go.mod, go.sum, files and directories with the prefix ".", files with the suffix ".md" or capitalized files name are allowed); found existing file "go1.18.1.linux-amd64.tar.gz"
```

**Comment by camilamacedo86** [comment]
Hi @viveksahu26, 

The 1.18 is not supported so far. So, trying to use the latest Kubebuilder release with go 1.18 will fail.

**Comment by camilamacedo86** [comment]
This issue was discussed in the C+R, Kubebuilder meeting yesterday. @AlmogBaku, thank you for helping us identify the impact and how we can address this requirement.  

@varshaprasad96 raised a good point and check that we had been only updating the go version to officially support when we update the deps to avoid issues faced in the past. She could identify some scenarios such as: 

> For example, with some recent changes, we can't use k8s 1.23 with older versions of Go (For ex: https://github.com/operator-framework/operator-sdk/issues/5558#issuecomment-1047137754) - this is a vice versa case. 

Therefore, for we are able to bump the go version 1.18 for the go/v3 plugin:

a) Update controller runtime (Its maintainers are ok with us doing that now in the master since it is using k8s 1.24 alpha deps already)
b) Waiting for a controller-runtime new release using go 1.18
c) Bump go 1.18 and controller-runtime on Kubebuilder go/v3 plugin and CLI

Then, to update the Kubebuilder CLI/API:

The config-gen ( which is an alpha, experimental and optional option ) uses controller tools as dep importing the code. So we also need to:
a) Update controller tools with go 1.18 and its deps to use k8s 1.24
b)Then, ask for a controller-tools release 
c) So that, we can bump Kubebuilder CLI/API and deps to use go 1.18, controller-tools, and k8s api 1.24 
d) When the above PR be done we will need to also get merged the PR to update the infra tests: https://github.com/kubernetes/test-infra/pull/25775

Additionally: we should not update go/v2 

I am adding this info to the description so that we can clarify what needs to be done here so we are able to accomplish this goal. 

c/c @ryanbrainard @rashmigottipati

**Comment by AlmogBaku** [comment]
I think go/v2 should also use go 1.18; there is no reason not to do so :)

**Comment by camilamacedo86** [comment]
Hi @AlmogBaku, 

We can only update go/v2 to use go version 1.18 if we would be able to upgrade the controller runtime and k8s deps used on it. Then, if we want to do that we need to check if has any reason for we have been not updating the go/v2 first. ( I believe that we have )

**Comment by camilamacedo86** [comment]
Hi @AlmogBaku, 

We can only update go/v2 to use go version 1.18 if we would be able to upgrade the controller runtime and k8s deps used on it. Then, if we want to do that we need to check if has any reason for we have been not updating the go/v2 first.

**Comment by camilamacedo86** [comment]
Hi @AlmogBaku, 

We will also need to update and change the release tools to work with go 1.18
- Update the etcd and Kubernetes versions with 1.24 also golang : https://github.com/kubernetes-sigs/kubebuilder/tree/tools-releases/build
- We also need to update the project : https://github.com/kubernetes-sigs/kubebuilder-release-tools: (we need to make the release notes work with go 1.18 and we need to be able to install it as a module)

**Comment by Vaxuite** [comment]
@camilamacedo86  I've done the controller runtime pr https://github.com/kubernetes-sigs/controller-runtime/pull/1895/files. Ill have a look at controller-tools in the coming days.

**Comment by Vaxuite** [comment]
Are you sure we need to update controller-runtime @camilamacedo86 it already works with go 1.18?

**Comment by AlmogBaku** [comment]
Hi, As far as I know, we're waiting for controller-runtime and controller-tools releases

**Comment by zbindenren** [comment]
Just as information. I am developping an operator with go1.18 and k8s 1.24.0. I had problems with `controller-gen` which paniced. I fixed this issue with the following:

```console
$ go install -v sigs.k8s.io/controller-tools/cmd/controller-gen@master
$ cp $HOME/go/bin/controller-gen $PROJECT_PATH/bin/
```

**Comment by damdo** [comment]
@AlmogBaku looks like new releases for [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/releases/latest) and [controller-tools](https://github.com/kubernetes-sigs/controller-tools/releases/latest) are both out now.

**Comment by camilamacedo86** [comment]
We are blocked by: https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/pull/217 now. 
After this one is merged we are good to go.

**Comment by bruceherve** [comment]
Hi. for anyone struggling  with this issue you can scaffold by using the go/v2 ie run **kubebuilder init --plugins go/v2** as you can see,you are using the go/v2  instead of go/v3 plugin..and once you run **kubebuilder create api** there will be no issue even if you are using Go 1.81..So I guess this allows you to use **kubebuilder with Go 1.81** while waiting for the upgrades to be made..

**Comment by calvinpark** [comment]
I started a PR in the homebrew-core repo to fix the go version dependency: https://github.com/Homebrew/homebrew-core/pull/102791
I'm new to homebrew formula edits so any help would be appreciated.

**Comment by damdo** [comment]
https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/pull/217 has merged 🎉  cc. @camilamacedo86

**Comment by camilamacedo86** [comment]
From now on, master branch should support go 1.18

**Comment by camilamacedo86** [comment]
@qingfengfenga please, open a new issue with your specific problem and providing the details steps to allow us to reproduce it.

**Comment by qingfengfenga** [comment]
@camilamacedo86 Sorry, when I checked the version and prepared to reproduce it locally, I found that my document was not the latest (translated version in other languages). After I checked the latest document, using go1.18.1 can complete all functions of cronjob, except testing ( #2642  I found the description of the test problem here). Thank you.
